pkg/controller/garbagecollector/graph_builder.go

```diff
@@ -0,0 +1,501 @@
+	finalizers := accessor.GetFinalizers()
+	for _, finalizer := range finalizers {
+		if finalizer == v1.FinalizerDeleteDependents {
+			return true
+		}
+	}
+	return false
+}
+
+func hasOrphanFianlizer(accessor meta.Object) bool {
+	finalizers := accessor.GetFinalizers()
+	for _, finalizer := range finalizers {
+		if finalizer == v1.FinalizerOrphanDependents {
+			return true
+		}
+	}
+	return false
+}
+
+// this function takes newAccessor directly because the caller already
+// instantiates an accessor for the newObj.
+func startsWaitingForDependentsDeleted(oldObj interface{}, newAccessor meta.Object) bool {
+	return deletionStarts(oldObj, newAccessor) && hasDeleteDependentsFinalizer(newAccessor)
+}
+
+// this function takes newAccessor directly because the caller already
+// instantiates an accessor for the newObj.
+func startsWaitingForDependentsOrphaned(oldObj interface{}, newAccessor meta.Object) bool {
+	return deletionStarts(oldObj, newAccessor) && hasOrphanFianlizer(newAccessor)
+}
+
+// if an blocking ownerReference points to an object gets removed, or gets set to
+// "BlockOwnerDeletion=false", add the object to the dirty queue.
+func (gb *GraphBuilder) addUnblockedOwnersToDeleteQueue(removed []metav1.OwnerReference, changed []ownerRefPair) {
+	for _, ref := range removed {
+		if ref.BlockOwnerDeletion != nil && *ref.BlockOwnerDeletion {
+			node, found := gb.uidToNode.Read(ref.UID)
+			if !found {
+				glog.V(5).Infof("cannot find %s in uidToNode", ref.UID)
+				continue
+			}
+			gb.attemptToDelete.Add(node)
+		}
+	}
+	for _, c := range changed {
+		wasBlocked := c.oldRef.BlockOwnerDeletion != nil && *c.oldRef.BlockOwnerDeletion
+		isUnblocked := c.newRef.BlockOwnerDeletion == nil || (c.newRef.BlockOwnerDeletion != nil && !*c.newRef.BlockOwnerDeletion)
+		if wasBlocked && isUnblocked {
+			node, found := gb.uidToNode.Read(c.newRef.UID)
+			if !found {
+				glog.V(5).Infof("cannot find %s in uidToNode", c.newRef.UID)
+				continue
+			}
+			gb.attemptToDelete.Add(node)
+		}
+	}
+}
+
+func (gb *GraphBuilder) processTransitions(oldObj interface{}, newAccessor meta.Object, n *node) {
+	if startsWaitingForDependentsOrphaned(oldObj, newAccessor) {
+		glog.V(5).Infof("add %s to the attemptToOrphan", n.identity)
+		gb.attemptToOrphan.Add(n)
+		return
+	}
+	if startsWaitingForDependentsDeleted(oldObj, newAccessor) {
+		glog.V(2).Infof("add %s to the attemptToDelete, because it's waiting for its dependents to be deleted", n.identity)
+		// if the n is added as a "virtual" node, its deletingDependents field is not properly set, so always set it here.
+		n.setDeletingDependents(true)
+		for dep := range n.dependents {
+			gb.attemptToDelete.Add(dep)
+		}
+		gb.attemptToDelete.Add(n)
+	}
+}
+
+func (gb *GraphBuilder) runProcessGraphChanges() {
+	for gb.processGraphChanges() {
+	}
+}
+
+// Dequeueing an event from graphChanges, updating graph, populating dirty_queue.
+func (gb *GraphBuilder) processGraphChanges() bool {
```

Comment:
@lavalamp this function is adapted from [processEvent](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/garbagecollector/garbagecollector.go#L347), which builds the dependency graph, adds object to attemptToOrphan or attemptToDelete queue. This function updated to handle objects that need foreground GC.